File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/map/overlay/overlay/LazyOverlay.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay

import dev.note11.flutter_naver_map.flutter_naver_map.model.enum.NOverlayType
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.info.NOverlayInfo

internal sealed interface LazyOverlay<NO : Any, O : AddableOverlay<*>> : LazyOrAddableOverlay {
    val info: NOverlayInfo
    abstract override fun createMapOverlay(): NO

    val wrappedMarker: O

    companion object {
        fun fromMessageable(
            info: NOverlayInfo,
            args: Map<String, Any>,
        ): LazyOverlay<*, *> {
            require(info.type.isLazy)

            return when (info.type) {
                NOverlayType.CLUSTERABLE_MARKER -> NClusterableMarker.fromMessageable(args)
                else -> throw IllegalArgumentException("${info.type} is not LazyOverlay")
            }
        }
    }
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/model/map/overlay/overlay/LazyOverlay.swift
import NMapsMap

internal protocol LazyOverlay: LazyOrAddableOverlay {
    override associatedtype OverlayType: Any
    associatedtype RealOverlayType: AddableOverlay // Wrapped
    
    override func createMapOverlay() -> OverlayType
    
    var info: NOverlayInfo { get }
    var wrappedOverlay: RealOverlayType { get }
}

internal func asLazyOverlayFromMessageable(info: NOverlayInfo, args: Dictionary<String, Any>) throws -> any LazyOverlay  {
    return switch info.type {
    case .clusterableMarker: NClusterableMarker.fromMessageable(args)
    default: throw NSError(domain: "LazyOverlay", code: 0, userInfo: [
        NSLocalizedDescriptionKey: "\(info.type) is not LazyOverlay"
    ])
    }
}
